### PR TITLE
firstorder: default tactic is “auto with core”

### LINKED ISSUE #7724: The default for Firstorder Solver should be auto, not auto with *
#### Version

8.8.0

#### Description of the problem

It's very unfortunate that `firstorder` means `firstorder auto with *` by default.  I think we should deprecate this behavior by setting it to `auto` by default and adding `Global Set Firstorder Solver auto with *` to the compat file.  (And similarly for `intuition`.)

### PR DESCRIPTION
**Pull request by vbgl**
The benchmarks of #11018 show that using `lia` instead of `omega` in “auto with zarith” incur some performance penalty.

Conjecture: some time is wasted by `lia` in trying hard to solve some goals without success. This PR attempts at removing some calls to “auto with zarith” that may be unexpected.

Currently, the `firstorder` tactics ~(including `intuition`)~ call “auto with *” so discharge leaf goals. This PR restricts the hint database to `core`.

Fixes #7724

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Overlays:
 - [x] https://github.com/mit-plv/kami/pull/17
 - [x] https://github.com/coq-community/aac-tactics/pull/55
 - [x] https://github.com/mit-plv/fiat/pull/34
 - [x] https://github.com/damien-pous/relation-algebra/pull/11
 - [x] https://github.com/Karmaki/coq-dpdgraph/pull/71
 - [x] https://github.com/PrincetonUniversity/VST/pull/390
 - [x] https://github.com/fblanqui/color/pull/21

### CODE REVIEW

**Comment by maximedenes** [comment]
I think it is fairly consensual that it would indeed be better to restrict the set of databases used. What would be the plan to make the transition not too painful on the user side, though? Is there a way to emit a warning each time a development relies on `firstorder` calling `auto with *` rather than `auto with core`?

**Comment by Zimmi48** [comment]
> Is there a way to emit a warning each time a development relies on `firstorder` calling `auto with *` rather than `auto with core`?

It seems feasible to collect the list of lemmas that were used and which hint database they came from. This would be a nice feature to have more generally speaking as it could be used to suggest the user to replace a call to `auto with *` with `auto with <list of hint dbs>` or even `auto using <list of lemmas>` when the list of lemmas that were used is short.

**Comment by vbgl** [comment]
I did not try, but I guess that you can set the `Firstorder Solver` to something like “auto with core || info_auto with *” and get useful traces.

**Comment by fakusb** [comment]
> It seems feasible to collect the list of lemmas that were used and which hint database they came from.
This probably falls short when using extern hints.

File: theories/FSets/FSetBridge.v
Comment (JasonGross) [inline theories/FSets/FSetBridge.v]: Please let's take this opportunity to get rid of `auto with *` in the standard library.  Since https://github.com/coq/coq/commit/55699e2b9a91356b7d43c4096f65fb199777b9a1 , `debug auto with *` prints which databases the hints come from, which should make it relatively easy to know what databases to use explicitly.

File: theories/Lists/List.v
Comment (JasonGross) [inline theories/Lists/List.v]: Idem

File: theories/MSets/MSetPositive.v
Comment (JasonGross) [inline theories/MSets/MSetPositive.v]: Idem

File: theories/Numbers/Cyclic/Abstract/NZCyclic.v
Comment (JasonGross) [inline theories/Numbers/Cyclic/Abstract/NZCyclic.v]: Idem

File: theories/Structures/EqualitiesFacts.v
Comment (JasonGross) [inline theories/Structures/EqualitiesFacts.v]: Idem

**Review by JasonGross — CHANGES_REQUESTED** [review]
This is a step in the right direction!  I think we should take this opportunity to purge the standard library of `auto with *`.

Please also add a changelog entry and update the relevant Compat file to `Set Firstorder Solver auto with *`.

**Comment by JasonGross** [comment]
Does `info_auto with *` even work?  Note that https://github.com/coq/coq/commit/55699e2b9a91356b7d43c4096f65fb199777b9a1 added database information to `debug auto`, and the issue where I requested that feature mentioned that `info_auto with *` is broken (I cannot recall if I ever reported a separate issue for that)

**Comment by vbgl** [comment]
Thanks for the comments. I’ve made sure not to introduce new `auto with *`, updated the `Compat.Coq812` module and added a changelog entry.

`info_auto with *` works (displays a trace including the databases that are used) but `info_eauto with *` does not.

In an additional commit, I’ve removed from the standard library a few calls to `auto with *`. Many more are left but I don’t intend to do more of such cleaning right now.

**Review by Zimmi48 — APPROVED** [review]
LGTM!

It would be nice to mention in the changelog the trick that `debug auto with *` shows which databases were used, the option that allows to change the default firstorder tactic, and the fact that the `-compat 8.12` option resets this option to `auto with *`.

**Comment by vbgl** [comment]
I’ve updated the changelog as suggested by Théo. The overlay for CoLoR remains to be merged. Otherwise, I think this is ready.

Latest benchmark look good:

~~~
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │
│                             │                         │                                             │                                             │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                    coq-corn │ 1383.12 1426.12 -3.02 % │     3862774148559     3977814191112 -2.89 % │     6010866267832     6159765996222 -2.42 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                  coq-sf-plf │   42.01   42.36 -0.83 % │      117504510904      117672108681 -0.14 % │      159665026390      159668829478 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│            coq-math-classes │  204.00  205.59 -0.77 % │      569145091802      574168568953 -0.87 % │      764968884209      770361656290 -0.70 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                coq-coqprime │   84.36   84.89 -0.62 % │      232369625348      232324726215 +0.02 % │      333962264829      333608032099 +0.11 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│      coq-mathcomp-ssreflect │   49.28   49.51 -0.46 % │      135166283359      135300452389 -0.10 % │      176592138331      176576835325 +0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                coq-bedrock2 │  200.90  201.58 -0.34 % │      556861579607      557225724297 -0.07 % │      929987620007      929835320016 +0.02 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│              coq-verdi-raft │ 1247.38 1250.62 -0.26 % │     3481515465818     3485995230889 -0.13 % │     4912418461555     4912425139700 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                    coq-hott │  303.20  303.92 -0.24 % │      822301488524      822994189896 -0.08 % │     1322272282922     1322356745838 -0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                coq-compcert │  672.64  673.49 -0.13 % │     1874065284330     1878409265451 -0.23 % │     2630533204608     2630450781323 +0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                  coq-geocoq │ 1526.14 1527.04 -0.06 % │     4259553443899     4261579108183 -0.05 % │     6495893280115     6496403199256 -0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                 coq-coqutil │   57.07   57.10 -0.05 % │      158456627914      158136484865 +0.20 % │      202404633155      202264456193 +0.07 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│          coq-mathcomp-field │  262.48  262.60 -0.05 % │      730935392580      731180491132 -0.03 % │     1124351032330     1124133121522 +0.02 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│      coq-mathcomp-character │  163.73  163.80 -0.04 % │      456133876782      456312695742 -0.04 % │      639344865270      639335120674 +0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│      coq-mathcomp-odd-order │ 1357.30 1357.72 -0.03 % │     3782189896670     3785677903385 -0.09 % │     6726566759614     6726585321079 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                   coq-flocq │  164.33  164.28 +0.03 % │      456005437798      456041388002 -0.01 % │      585935895239      585882012315 +0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│               coq-fourcolor │ 2091.95 2091.10 +0.04 % │     5836374653563     5835864297917 +0.01 % │    11466045758275    11465582788114 +0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│             coq-lambda-rust │ 1539.07 1538.32 +0.05 % │     4289293376747     4287334759171 +0.05 % │     6174840985629     6170539557725 +0.07 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                 coq-unimath │ 4031.38 4028.51 +0.07 % │    11228628726124    11216410518625 +0.11 % │    21358666709123    21364364133432 -0.03 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                   coq-verdi │  115.73  115.58 +0.13 % │      319002674687      319649762991 -0.20 % │      408526425515      408715735475 -0.05 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│             coq-fiat-crypto │ 4431.02 4425.10 +0.13 % │    12311198355882    12292927938202 +0.15 % │    20338412656780    20338437045786 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│        coq-mathcomp-algebra │  145.85  145.45 +0.28 % │      405586307798      405631181001 -0.01 % │      539718188118      539729115804 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│       coq-mathcomp-solvable │  176.03  175.44 +0.34 % │      489751303070      488980219722 +0.16 % │      697565659505      697571279074 -0.00 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                 coq-bignums │   62.61   62.40 +0.34 % │      173965611776      173645113840 +0.18 % │      243683413495      243708543239 -0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│            coq-fiat-parsers │  599.46  597.33 +0.36 % │     1678000798319     1673004507285 +0.30 % │     2674148157121     2673979929905 +0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│ coq-rewriter-perf-SuperFast │ 1002.56  996.12 +0.65 % │     2791284816291     2776452333934 +0.53 % │     4545302748641     4544700893529 +0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│       coq-performance-tests │   11.58   11.50 +0.70 % │       31635892224       31569789965 +0.21 % │       54782695981       54763663427 +0.03 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│               coq-fiat-core │  102.89  102.10 +0.77 % │      290614435489      290083136748 +0.18 % │      386769764882      386562181150 +0.05 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│                coq-rewriter │  580.72  576.18 +0.79 % │     1613171555132     1599356901417 +0.86 % │     2440791258864     2440543710570 +0.01 % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┤
│       coq-mathcomp-fingroup │   49.48   49.08 +0.81 % │      136923808651      136471886657 +0.33 % │      190277816992      190275228138 +0.00 % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┘
~~~

File: doc/changelog/04-tactics/11760-firstorder-leaf.rst
Comment (JasonGross) [inline doc/changelog/04-tactics/11760-firstorder-leaf.rst]: This is misleading.  There is no option to set the default solver for intuition, and the compat flag does not impact it.  I think it'd be nice if we added a flag for intuition like there is one for firstorder

**Review by JasonGross — CHANGES_REQUESTED** [review]

File: doc/changelog/04-tactics/11760-firstorder-leaf.rst
Comment (vbgl) [reply doc/changelog/04-tactics/11760-firstorder-leaf.rst]: Indeed. This PR does not change the behavior of `intuition`. I’ve thus fixed the changelog.

**Comment by vbgl** [comment]
Jason, are you satisfied with my last changes?

May someone want to take care of merging this PR?

**Comment by Zimmi48** [comment]
I can take this since @coq/firstorder-maintainers didn't. We need to wait for the last overlay to be merged though.

File: doc/changelog/04-tactics/11760-firstorder-leaf.rst
Comment (JasonGross) [inline doc/changelog/04-tactics/11760-firstorder-leaf.rst]:
```suggestion
  to ease the migration of legacy code, the default solver can be set to `debug auto with *`
  with `Set Firstorder Solver debug auto with *`
```

**Review by JasonGross — APPROVED** [review]
I have edited the initial PR comment to add "Fixes #7724".  I would like to see a similar change for `intuition` (#4949 and #7725), but that can happen in another PR.

I've suggested a change to the changelog entry that I think will make it easier for users who are unfamiliar with this machinery to do the migration (it eliminates the step of them needing to read the section of the reference manual).  Feel free to merge with or without this change, though obviously I'd prefer it in.

**Comment by Zimmi48** [comment]
Bedrock2 failure is an OOM which also occurs in master.

**Comment by vbgl** [comment]
Indeed. bedrock2 builds fine locally and on roquableu.

**Comment by RalfJung** [comment]
This most likely caused a regression: https://github.com/coq/coq/issues/11865

**Comment by Zimmi48** [comment]
Does not look like a regression to me.

**Comment by RalfJung** [comment]
Well it's a proof that worked before but fails now, which by definition is a regression. ;)

It just seems that in this case, that was deliberate -- fair.

**Comment by Zimmi48** [comment]
> Well it's a proof that worked before but fails now, which by definition is a regression. ;)

Nope. *By definition*, a regression is when the change was not deliberate.

**Pull request merged**
